### PR TITLE
add coherence future-compat warnings for marker-only trait objects

### PR DESCRIPTION
**Pull request by arielb1**
The future-compat warnings break code that assumes that `dyn Send + Sync !=
dyn Sync + Send`, and are the first step in making them equal. cc #33140.

Note: this lint should be made to default-warn before we merge. It is deny only for the crater run.

r? @nikomatsakis / @scalexm . cc @Centril & @alexreg.

### CODE REVIEW

File: src/test/run-pass/issues/issue-33140.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
```
Comment (Centril) [inline src/test/run-pass/issues/issue-33140.rs]: Can you move this file over (after crater is done...) to the ui tests? You can add `// run-pass` for that in the ui test file.

Also, you don't need to add copyright headers anymore. :)

Comment (arielb1) [reply src/test/run-pass/issues/issue-33140.rs]: sure

**Comment by arielb1** [comment]
please don't try this until travis is green!

**Comment by kennytm** [comment]
@craterbot run start=master#c8ae2de8363be088604ffa2ccba84ff7ad7250bc end=try#1df97983eb2057dc6a8f7eb364e263b51ae23230 mode=check-only

**Comment by alexreg** [comment]
That traitobject crate was an issue for my duplicate auto traits lint too.

File: src/librustc/lint/builtin.rs [hunk 2]
```diff
@@ -208,6 +208,13 @@ declare_lint! {
     "potentially-conflicting impls were erroneously allowed"
 }
 
+declare_lint! {
+    pub ORDER_DEPENDENT_TRAIT_OBJECTS,
+    Deny,
+    // FIXME: ENG
```
Comment (nikomatsakis) [inline src/librustc/lint/builtin.rs]: What does this FIXME refer to?

**Review by nikomatsakis — APPROVED** [review]
Code seems ok for now, apart from @Centril's nit

**Comment by nikomatsakis** [comment]
huh, so [the `loggr` regression](https://crater-reports.s3.amazonaws.com/pr-56481/try%231df97983eb2057dc6a8f7eb364e263b51ae23230/reg/loggr-0.3.0/log.txt) from the crater report does not look related:

```
Dec 05 10:49:36.575 INFO [stderr] error[E0599]: no associated item named `O_CREAT` found for type `unix::file::OpenFlags` in the current scope
Dec 05 10:49:36.575 INFO [stderr]   --> /cargo-home/registry/src/github.com-1ecc6299db9ec823/unix-0.5.5/src/file.rs:44:35
Dec 05 10:49:36.575 INFO [stderr]    |
Dec 05 10:49:36.575 INFO [stderr] 44 | pub const O_CREAT   : OpenFlags = OpenFlags::O_CREAT;
Dec 05 10:49:36.575 INFO [stderr]    |                                   ^^^^^^^^^^^^^^^^^^ associated item not found in `unix::file::OpenFlags`
```

**Comment by Centril** [comment]
@nikomatsakis we got the same "regression" in https://github.com/rust-lang/rust/pull/51285#issuecomment-446159507 so I think it is spurious.

**Comment by nikomatsakis** [comment]
@rust-lang/lang meeting says :shipit: -- seems like we should clearly go forward,.

**Comment by nikomatsakis** [comment]
I guess for now we can land this version, it'd be nice to do at least a *short* warning period, though I'd definitely like to land https://github.com/rust-lang/rust/pull/56837 ASAP

**Comment by Centril** [comment]
It seems we went for deny-by-default here instead of warn; was that intentional? cc @nikomatsakis

**Comment by arielb1** [comment]
It wasn't intentional - I forgot to change it to warn. However, I think the only affected crate is `traitobject`, so I think this is OK.

**Comment by Centril** [comment]
@arielb1 My concern isn't so much `traitobject` as the reverse dependencies it has (see https://github.com/rust-lang/rust/pull/55994#issuecomment-444308349 ...). Ideally we'd like to get a patch release of `traitobject` ASAP but we haven't been able to reach @reem yet.

**Comment by arielb1** [comment]
Having a deny-by-default lint shouldn't affect reverse dependencies.

**Comment by alexreg** [comment]
@arielb1 Okay. Should I change it to warn-by-default anyway?

**Pull request merged**